File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/utils/RequestBuilderUtils.java
package com.appcoins.sdk.billing.utils;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;
import java.util.Map;

public class RequestBuilderUtils {

  public static String buildUrl(String baseUrl, String endPoint, List<String> paths,
      Map<String, String> queries) {
    boolean hasQueries = !queries.isEmpty();
    if (endPoint == null) {
      endPoint = "";
    }
    StringBuilder urlBuilder = new StringBuilder(baseUrl + endPoint);
    for (String path : paths) {
      if (path != null) {
        buildPath(path, urlBuilder);
      }
    }
    if (hasQueries) {
      urlBuilder.append("?");
    }
    for (Map.Entry<String, String> entry : queries.entrySet()) {
      if (entry.getValue() != null && entry.getKey() != null) {
        buildQuery(entry, urlBuilder);
      }
    }
    if (hasQueries) {
      urlBuilder.deleteCharAt(urlBuilder.length() - 1);
    }
    return urlBuilder.toString();
  }

  private static void buildQuery(Map.Entry<String, String> entry, StringBuilder urlBuilder) {
    String key = "";
    String value = "";
    try {
      key = URLEncoder.encode(entry.getKey(), "utf-8");
      value = URLEncoder.encode(entry.getValue(), "utf-8");
    } catch (UnsupportedEncodingException e) {
      e.printStackTrace();
    }
    urlBuilder.append(key).append("=").append(value).append("&");
  }

  private static void buildPath(String path, StringBuilder urlBuilder) {
    String encodedPath = "";
    try {
      encodedPath = URLEncoder.encode(path, "utf-8");
    } catch (UnsupportedEncodingException e) {
      e.printStackTrace();
    }
    //Encoder transforms "=" into %3D, but it's not needed for paths
    encodedPath = encodedPath.replaceAll("%3D", "=");
    urlBuilder.append("/")
        .append(encodedPath);
  }

  public static String buildBody(Map<String, Object> bodyKeys) {
    StringBuilder builder = new StringBuilder("{");
    if (bodyKeys != null) {
      for (Map.Entry<String, Object> entry : bodyKeys.entrySet()) {
        if (entry.getValue() != null) {
          String value = entry.getValue()
              .toString();
          if (isString(entry.getValue())) {
            value = "\"" + value + "\"";
          }
          if (isMap(entry.getValue())) {
            value = buildBody((Map) entry.getValue());
          }
          builder.append("\"").append(entry.getKey()).append("\"").append(":").append(value).append(",");
        }
      }
      if (!bodyKeys.isEmpty()) {
        builder.deleteCharAt(builder.length() - 1);
      }
    }
    builder.append("}");
    return builder.toString();
  }

  private static boolean isMap(Object value) {
    return value instanceof Map;
  }

  private static boolean isString(Object value) {
    return value instanceof String && !((String) value).contains("{") && !((String) value).contains(
        "[");
  }
}

File: appcoins-in-game-updates/src/main/java/com/appcoins/sdk/ingameupdates/utils/RequestBuilderUtils.kt
package com.appcoins.sdk.ingameupdates.utils

import java.io.UnsupportedEncodingException
import java.net.URLEncoder

object RequestBuilderUtils {

    @JvmStatic
    fun buildUrl(
        baseUrl: String,
        endPointExternal: String?,
        paths: List<String?>,
        queries: Map<String?, String?>
    ): String {
        var endPoint = endPointExternal
        val hasQueries = queries.isNotEmpty()
        if (endPoint == null) {
            endPoint = ""
        }
        val urlBuilder = StringBuilder(baseUrl + endPoint)
        for (path in paths) {
            if (path != null) {
                buildPath(path, urlBuilder)
            }
        }
        if (hasQueries) {
            urlBuilder.append("?")
        }
        for (entry in queries.entries) {
            if (entry.value != null && entry.key != null) {
                buildQuery(entry, urlBuilder)
            }
        }
        if (hasQueries) {
            urlBuilder.deleteCharAt(urlBuilder.length - 1)
        }
        return urlBuilder.toString()
    }

    private fun buildQuery(entry: Map.Entry<String?, String?>, urlBuilder: StringBuilder) {
        var key = ""
        var value = ""
        try {
            key = URLEncoder.encode(entry.key, "utf-8")
            value = URLEncoder.encode(entry.value, "utf-8")
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
        }
        urlBuilder.append("$key=$value")
            .append("&")
    }

    private fun buildPath(path: String, urlBuilder: StringBuilder) {
        var encodedPath = ""
        try {
            encodedPath = URLEncoder.encode(path, "utf-8")
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
        }
        //Encoder transforms "=" into %3D, but it's not needed for paths
        encodedPath = encodedPath.replace("%3D".toRegex(), "=")
        urlBuilder.append("/")
            .append(encodedPath)
    }

    @JvmStatic
    fun buildBody(bodyKeys: Map<String, Any?>?): String {
        val builder = StringBuilder("{")
        if (bodyKeys != null) {
            for ((key, value1) in bodyKeys) {
                if (value1 != null) {
                    var value = value1.toString()
                    if (isString(value1)) {
                        value = "\"" + value + "\""
                    }
                    if (isMap(value1)) {
                        value = buildBody(value1 as Map<String, *>?)
                    }
                    builder.append("\"$key\":$value")
                        .append(",")
                }
            }
            if (bodyKeys.isNotEmpty()) {
                builder.deleteCharAt(builder.length - 1)
            }
        }
        builder.append("}")
        return builder.toString()
    }

    private fun isMap(value: Any?): Boolean {
        return value is Map<*, *>
    }

    private fun isString(value: Any?): Boolean {
        return value is String && !value.contains("{") && !value.contains(
            "["
        )
    }
}
